dotcom-rendering/src/server/lib/logging-middleware.ts

```diff
@@ -12,38 +18,56 @@ const hasPageId = (body: unknown): body is { pageId: string } => {
 };
 
 /**
- * An Express middleware which handles creating our logger store and logging requests after they've
- * completed.
+ * An Express middleware which wraps the inbuilt json() body-parser middleware to record execution time.
  */
-export const requestLoggerMiddleware: RequestHandler = (req, res, next) => {
+const expressJsonMiddleware = express.json();
+export const timedExpressJsonMiddleware: RequestHandler = (req, res, next) => {
+	recordTimeStart('json');
+
+	// Pass a wrapper around Express's next() callback which record when expressJsonMiddleware calls next()
+	expressJsonMiddleware(req, res, () => {
+		recordTimeStop('json');
+		next();
+	});
+};
+
+/**
+ * An Express middleware which handles creating a logger store to save information to be logged in the request logger.
+ */
+export const loggingStoreMiddleware: RequestHandler = (req, res, next) => {
 	const loggerState = {
 		request: {
-			pageId: hasPageId(req.body) ? req.body.pageId : 'no-page-id-found',
+			pageId: 'before-request-logger-middleware',
 			path: req.path,
 			method: req.method,
 		},
-		timing: {},
+		timing: {
+			total: Date.now(),
+		},
 	};
 
-	res.on('finish', () => {
-		const { request, error } = loggingStore.getStore() ?? {};
+	loggingStore.run(loggerState, () => {
+		next();
+	});
+};
 
-		if (!request?.type) return;
+/**
+ * An Express middleware which handles logging requests and their associated details after they have been completed.
+ */
+export const requestLoggerMiddleware: RequestHandler = (req, res, next) => {
+	// We don't have access to the page ID until after the express.json() middleware runs.
+	recordPageId(hasPageId(req.body) ? req.body.pageId : 'no-page-id-found');
+
+	res.on('finish', () => {
+		recordTimeStop('total');
```

Comment:
Why are we not explicitly set `recordTimeStart('total')` as well in the flow (not here)? Are we inferring it from another `RecordablePeriod`?

Reply:
Good question!

The reason I don't need to invoke recordTimeStart('total') is because it gets initialized when the logging store is constructed in order to avoid any delay to the total time.

https://github.com/guardian/dotcom-rendering/pull/7681/files#diff-d24c7acadf933f1da8ff9460b4fef129f48c318a3290aa4a87253128b425a8fdR45